agent/loop.py

```
@@ -577,6 +577,15 @@ async def run(
             if key_id:
                 self.key_id = key_id
 
+            # ── Learning loop: persist the cause of every failed step so the
+            # next run's planner sees it (agent/lessons.py). Retry without
+            # this is not learning — the same mistake recurs forever.
+            try:
+                from agent.lessons import record_step_failures
+                record_step_failures(plan.goal, step_results)
+            except Exception:
+                pass
```

Ruff (0.15.20)

[error] `try`-`except`-`pass` detected, consider logging the exception

(S110)